src/lib/common/include/sol-mainloop.h

```diff
@@ -64,11 +64,269 @@ extern "C" {
+ * @endcode
+ *
+ * Note that the @c rx_cb returns a @c ssize_t, the returned value must be @c -errno if an error happened or the number
+ * of bytes consumed from @c buf (0 is valid). The consumed bytes will be removed from the @c buf.
+ *
+ * Below there's an example of the read implementation:
+ *
+ * @code{.c}
+ * //my_stream_api.c
+ *
+ * struct my_stream_api_handle {
+ *    Device *dev;
+ *    const void *user_data;
+ *    void (*tx_cb)(void *user_data, my_stream_api_handle *handle, struct sol_blob *blob, int status);
+ *    ssize_t (*rx_cb)(void *user_data, my_stream_api_handle *handle, const struct sol_buffer *buf);
+ *    struct sol_ptr_vector pending_blobs;
+ *    size_t tx_size;
+ *    size_t pending_bytes;
+ *    struct sol_timeout *read_timeout;
+ * };
+ *
+ * //Code...
+ *
+ * static bool
+ * _inform_user(void *data)
+ * {
+ *    struct my_stream_api_handle *handle = data;
+ *    ssize_t r;
+ *
+ *    //Inform the user
+ *    r = handle->rx_cb((void *)handle->user_data, handle, handle->rx);
+ *    //Remove the data.
+ *    if (r < 0)
+ *       SOL_ERR("Something wrong happened %zd", r);
+ *   else
+ *       sol_buffer_remove_data(&uart->rx, 0, r);
+ *   if (handle->rx.used)
+ *      return true;
+ *   handle->read_timeout = NULL;
+ *   return false;
+ * }
+ * static void
+ * _can_read(void *data)
+ * {
+ *    struct my_stream_api_handle *handle = data;
+ *    size_r read;
+ *    ssize_t r;
+ *
+ *    //Append data to the buffer
+ *    read = my_device_read(handle->dev, sol_buffer_at_end(&uart->rx), uart->rx.capacity - uart->rx.used);
+ *    uart->rx.used += read;
+ *    //Keep reading until there's no more data
+ *    if (!handle->read_timeout)
+ *       handle->read_timeout = sol_timeout_add(0, _inform_user, handle);
+ * }
+ *
+ *
+ * struct my_stream_api_handle *
+ * my_stream_api_new(struct my_stream_api_config *config)
+ * {
+ *    struct my_stream_api_handle *handle;
+ *    size_t rx_size;
+ *    void *buf;
+ *
+ *    //Init stuff...
+ *
+ *    //The user does not want to read from the stream, ignore rx configuration.
+ *    if (!config->rx_cb)
+ *       return handle;
+ *
+ *    handle->rx_cb = config->rx_cb;
+ *    rx_size = config->rx_size;
+ *
+ *    //Use a default size.
+ *    if (!rx_size)
+ *       rx_size = DEFAULT_SIZE:
+ *
+ *    buf = malloc(rx_size);
+ *    if (!buf)
+ *        goto clean_up;
+ *
+ *    sol_buffer_init_flags(&uart->rx, buf, rx_size, SOL_BUFFER_FLAGS_FIXED_CAPACITY | SOL_BUFFER_FLAGS_NO_NUL_BYTE);
+ *    //Register input observer
+ *    my_device_inform_data_available_for_read(handle->dev, _can_read, handle);
+ *    return handle;
+ * clean_up:
+ *   //Clean up code...
+ * }
+ * //More code...
+ * @endcode
+ *
+ * @subsection implementations Implementations
+ *
+ * Soletta currently implements streams for the following modules:
+ * <ul>
+ * <li> @ref UART
+ *    <ul>
+ *    <li> sol_uart_write()
+ *    <li> @ref sol_uart_config
```

Comment:
I believe if you write `struct sol_uart_config`, that is going to auto-link. Please test.